### PR TITLE
fix: avoid missed logs when streaming startup logs

### PR DESCRIPTION
**Pull request by mafredri**
I've been seeing inconsistent behavior with the startup log streaming and while looking at the implementation, I noticed a few issues:

- There's a delay between fetching initial logs and subscribing to the pubsub for updates, this can result in lost messages
- We did not have an indicator for "end of log" meaning to accurately discern it, we'd have to subscribe to both startup script and workspace updates and get an up-to-date lifecycle state of the workspace
- We were relying on pubsub telling us when "end of log" is reached, but this message was never sent

With this refactor, we are not dependent on state from the pubsub and the agent tries to log an EOF at the end.

Thoughts:

- Currently we don't do anything special with the EOF message, but it could be used to e.g. show exit status as a final log or something similar

Here's an image where the script has completed, but logs are missing (quite a lot of logs):
image

[test]

### CODE REVIEW

**Review by mtojek — APPROVED** [review]
It is a relatively large PR, but as long as it's consistent with other logs properties (Level, Output, CreatedAt), it is 👍 from me.

**Comment by mafredri** [comment]
I will add, I'm not super happy with the DB change to facilitate the EOF. Ultimately I'd like to update the query to be like `SELECT array_agg(logs.*) AS logs, (agent.lifecycle not starting) AS eof`, but that is dependent on functionality not in sqlc: https://github.com/kyleconroy/sqlc/issues/2163

Open to alternatives.

The main thing IMO is that we want the log fetching query to be able to signal EOF so that consumers know what's up.

File: codersdk/workspaceagents.go [hunk 1]
```diff
@@ -543,8 +553,8 @@ func (c *Client) WorkspaceAgentStartupLogsAfter(ctx context.Context, agentID uui
 		defer close(closed)
 		defer close(logChunks)
 		defer conn.Close(websocket.StatusGoingAway, "")
-		var logs []WorkspaceAgentStartupLog
 		for {
+			var logs []WorkspaceAgentStartupLog
```
Comment (mafredri) [inline codersdk/workspaceagents.go:557]: This was a gnarly bug, had me scratching my head for a while 😬 (new test cases caught it)

**Comment by mafredri** [comment]
I didn't like how I left some edge cases in the implementation, so I've tried to address those now with the last few commits.

- And EOF can be sent by the agent, or it will be forcefully closed via agent lifecycle state change (agents choice)
- The EOF log entry is sent to consumers and filtered out in the WebUI
- Fixed a memory bug the in sdk (currently only used by tests, but will be used by ssh)

File: coderd/workspaceagents.go [hunk 2]
```diff
@@ -260,9 +260,15 @@ func (api *API) patchWorkspaceAgentStartupLogs(rw http.ResponseWriter, r *http.R
...
-	var endLog agentsdk.StartupLog
 	outputLength := 0
-	for _, logEntry := range req.Logs {
+	for i, logEntry := range req.Logs {
+		if logEntry.EOF && i != len(req.Logs)-1 {
+			httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+				Message: "EOF log must be the last log entry.",
+			})
+			return
+		}
```
Comment (mtojek) [inline coderd/workspaceagents.go:270]: nit: not sure if it's worth it, but if you move this check before for-loop, you could fail fast

File: coderd/workspaceagents.go [hunk 3]
```diff
@@ -412,6 +444,9 @@ func (api *API) workspaceAgentStartupLogs(rw http.ResponseWriter, r *http.Reques
 			return
 		}
 	}
+	if after < 0 {
```
Comment (mtojek) [inline coderd/workspaceagents.go:447]: Under which circumstances `after` is < 0?

File: site/src/testHelpers/entities.ts [hunk 4]
```diff
@@ -1649,25 +1649,36 @@ export const MockStartupLogs: TypesGen.WorkspaceAgentStartupLog[] = [
...
     output: "Installing v4.8.3 of the amd64 release from GitHub.",
     level: "info",
+    eof: false,
+  },
+  {
+    id: 166667,
+    created_at: "2023-05-04T11:30:42.599999Z",
+    output: "",
+    level: "info",
+    eof: true,
```
Comment (mtojek) [inline site/src/testHelpers/entities.ts:1681]: Thinking loud:

Let's say that due to a bug we won't insert/store the EOF in the database. Will it affect the presentation layer or coderd logic?

**Review by mtojek — APPROVED** [review]
👍

File: coderd/workspaceagents.go (hunk 2 above)
Comment (mafredri) [reply coderd/workspaceagents.go:270]: If we do that, we need to do two iterations on the logs, first to check if any has EOF and then to construct the new format?

File: coderd/workspaceagents.go (hunk 3 above)
Comment (mafredri) [reply coderd/workspaceagents.go:447]: Perhaps this should return an error, test cases were written with `-500`, and I didn't know the reason. I can remove this check as it doesn't matter. Added it while trying to figure out the memory bug in the sdk.

File: coderd/workspaceagents.go (hunk 2 above)
Comment (mtojek) [reply coderd/workspaceagents.go:270]: Ok, I missed that. Let's leave it then.

File: site/src/testHelpers/entities.ts (hunk 4 above)
Comment (mafredri) [reply site/src/testHelpers/entities.ts:1681]: Nope, the frontend completely ignores EOF at the moment, but we could use it in the future to show e.g. when the log was closed. Or, perhaps which error code it exited with.

Likewise, the frontend mainly relies on the agent lifecycle state, so if the state changes the log will be hidden anyway.

We could make writing EOF in lifecycle state change mandatory to succeed, simply means the agent will keep trying to change the state until we can finally write EOF. (Currently we only warn log). That would take care of the edge case you mention.

Comment (mafredri) [reply site/src/testHelpers/entities.ts:1681]: This is now handled in [`eb7a7a3` (#8029)](https://github.com/coder/coder/pull/8029/commits/eb7a7a366589c5d13e78e0bdc7ab7077135c2d8d). If we fail to write EOF, the agent will simply retry the lifecycle state report. An up-to-date agent will close the log file before this, though.

File: coderd/workspaceagents.go (hunk 3 above)
Comment (mafredri) [reply coderd/workspaceagents.go]: Fixed in [`a1a40f3` (#8029)](https://github.com/coder/coder/pull/8029/commits/a1a40f37f72ee37ddb808ed7024815ad099b4cd0)

**Review by mtojek — APPROVED** [review]

**Pull request merged**